### PR TITLE
Add reduce_on_plateau LR scheduler to contrib directory.

### PR DESCRIPTION
**Pull request by vz415**
Following pull request #505 and issue #221 with @mtthss suggestions on changing where the scheduler is located (contrib) and using GradientTransformationExtraArgs API. Let me know if there's anything else that needs to be done to merge this. Cheers.

### CODE REVIEW

**Comment by fabianp** [comment]
Thanks for the contribution @vz415 ! There seem to be some minor issues regarding formatting that break the tests (see https://github.com/google-deepmind/optax/actions/runs/6830352294).

Other than that this looks good to me. Green light to merge once the tests pass @mtthss ?

**Comment by vz415** [comment]
@fabianp fixed the spacing formatting issue 🤦‍♂️ and everything should be ready to merge.

File: optax/contrib/prodigy.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ class ProdigyState(NamedTuple):
 def prodigy(
     learning_rate: base.ScalarOrSchedule = 0.1,
     betas: tuple[float, float] = (0.9, 0.999),
-    beta3: float | None = None,
+    beta3: float = None,
```
Comment (vroulet) [inline optax/contrib/prodigy.py:49]: Thanks for catching that! I fixed it, you'll need to merge with main.

**Comment by fabianp** [comment]
Thanks @vz415 for the changes! A couple more minor things and then I think we're ready to merge:

1. Is there any paper we could cite that describes this approach? If so, please add them under "References" on the docstring (see examples on _src/alias.py)
2. Please add this new method to the API reference (docs/api.rst)

**Comment by vz415** [comment]
Hi @fabianp , I've addressed the issues below.
1. There's no seminal paper on the technique but here's the pytorch [documentation](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html). I can add some papers I've seen use it if wanted.
2. I added it to the API reference. Let me know if I need to edit it.

File: docs/api.rst [hunk 2]
```diff
@@ -619,6 +619,7 @@ Schedules
 .. autofunction:: piecewise_constant_schedule
 .. autofunction:: piecewise_interpolate_schedule
 .. autofunction:: polynomial_schedule
+.. autofunction:: optax.contrib.reduce_on_plateau.reduce_on_plateau
```
Comment (fabianp) [inline docs/api.rst:622]:
```suggestion
.. autofunction:: optax.contrib.reduce_on_plateau
```

File: optax/contrib/reduce_on_plateau.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  cooldown:int
+
+
+def reduce_on_plateau(
+  reduce_factor: float,
+  patience: int,
+  min_improvement:float,
+  cooldown:int
+) -> base.GradientTransformationExtraArgs:
+  """ Args:
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:48]:
```suggestion
  """Reduce learning rate when a metric has stopped improving. 
  
  Models often benefit from reducing the learning once learning stagnates.
  This scheduler reads a metrics quantity and if no improvement is seen for
  a ‘patience’ number of epochs, the learning rate is reduced.
  
  Args:
```

File: optax/contrib/reduce_on_plateau.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  plateau_count: int
+  lr: float
+  cooldown_counter: int
+  cooldown:int
+
+
+def reduce_on_plateau(
+  reduce_factor: float,
+  patience: int,
+  min_improvement:float,
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:45]:
```suggestion
  min_improvement: float,
```

File: optax/contrib/reduce_on_plateau.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  cooldown:int
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:46]:
```suggestion
  cooldown: int
```

File: optax/contrib/reduce_on_plateau.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+) -> base.GradientTransformationExtraArgs:
+  """ Args:
+  reduce_factor: Factor by which the learning rate will be reduced. 
+      new_lr = lr * factor.
+  patience: Number of epochs with no improvement after which learning
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:51]: Why are these parameters in terms of number of epochs? Wouldn't it be more accurate to talk about number of iterations? It seems to me that the method has now knowledge of epochs, but keeps track of iterations instead

File: optax/contrib/reduce_on_plateau.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+  Models often benefit from reducing the learning once learning stagnates.
+  his scheduler reads a metrics quantity and if no improvement is seen for
+  a ‘patience’ number of epochs, the learning rate is reduced
+
+  Args:
+
+  reduce_factor: Factor by which the learning rate will be reduced. 
+      new_lr = lr * factor.
+  patience: Number of epochs with no improvement after which learning
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:58]:
```suggestion
  patience: Number of iterations with no improvement after which learning 
```

File: optax/contrib/reduce_on_plateau.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+  Args:
+
+  reduce_factor: Factor by which the learning rate will be reduced. 
+      new_lr = lr * factor.
+  patience: Number of epochs with no improvement after which learning 
+      rate will be reduced.
+  min_improvement: Threshold for measuring the new optimum, to only focus on 
+      significant changes.
+  cooldown: Number of epochs to wait before resuming normal operation
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:62]:
```suggestion
  cooldown: Number of iterations to wait before resuming normal operation 
```

File: optax/contrib/reduce_on_plateau.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  reduce_factor: float,
+  patience: int,
+  min_improvement: float,
+  cooldown: int
+) -> base.GradientTransformationExtraArgs:
+  """ Reduce learning rate when a metric has stopped improving. 
+
+  Models often benefit from reducing the learning once learning stagnates.
+  his scheduler reads a metrics quantity and if no improvement is seen for
+  a ‘patience’ number of epochs, the learning rate is reduced
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:52]:
```suggestion
  a ‘patience’ number of epochs, the learning rate is reduced.
```

File: optax/contrib/reduce_on_plateau.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    params=None,
+    extra_args=None,
+  ):
+    del params
+    if extra_args is None:
+      extra_args = {}
+    current_loss = extra_args.get("loss")
+
+    # Check if the current loss is the best so far
+
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:91]:
```suggestion
```

File: optax/contrib/reduce_on_plateau.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+class ReduceLROnPlateauState(NamedTuple):
+  """State for the ReduceLROnPlateau callback."""
+  reduce_factor: float
+  patience: int
+  min_improvement: float
+  best_loss: float
+  plateau_count: int
+  lr: float
+  cooldown_counter: int
+  cooldown:int
```
Comment (fabianp) [inline optax/contrib/reduce_on_plateau.py:39]:
```suggestion
  cooldown: int
```

**Comment by fabianp** [comment]
Is there a reason why the parameters reduce_factor, patience, etc. need to be passed to the state ReduceLROnPlateauState ?

**Comment by fabianp** [comment]
Also, in its current state, I think this method is not straightforward to use (at least it was not clear to me). Would you be willing add a docstring or create an example on how to combine this scaling with (say) SGD or Adam?

**Comment by vz415** [comment]
Thanks for the review and feedback! Answering your question: those parameters are passed to update the learning rate after every epoch. Those parameters could externalized via functools.partial but I think this format makes it easier to implement for users, although more implicit in how the updates are carried out. Let me know if this makes sense or if you have other concerns.

I'll add a docstring with an example on how to use. If that's not clear I can create an example.

**Comment by fabianp** [comment]
This now merged with some changes. Among other things, I changed some keyword arguments to match those of the pytorch implementation.

Other changes were done to fix some internal failures due to the fact that the internal pytype checks are a bit more stringent than the ones that run on the github actions.

If you feel that some of these changes are for the worst, please open another PR and propose modifications.

Regarding documentation, I mentioned earlier I think it is important to make an example on using this function. I created issue #679 to track progress on this

**Comment by fabianp** [comment]
thanks for the answer @vz415 (and for your contribution)! I saw your comment after I had submitted. In any case please don't hesitate to open another PR to edit the submitted code
